File: py/hockeyAPI.py
import requests, json, sys
import matplotlib.pyplot as plt
import numpy as np
import math
from collections import namedtuple


class HockeyAPI:

    def __init__(self):
        self._API = 'https://statsapi.web.nhl.com/api/v1/'
        self._NHL_ID = 133
        self._Teams = self.get_teams()

    def get_api_call_dict(self, parameter):
        resp = requests.get(self._API+parameter)
        resp.raise_for_status()
        parameter_resp = resp.json().get(parameter)
        member_dict = {}
        for member in parameter_resp:
            member_dict[member.get('id')] = member.get('name')
        return member_dict

    def get_teams(self):
        resp = requests.get(self._API+'teams')
        resp.raise_for_status()
        teams = resp.json().get('teams')
        team_dict = {}
        for team in teams:
            team_dict[team.get('id')] = self.json_to_obj(team)
        return team_dict

    def get_roster(self, team_id):
        resp = requests.get(self._API + 'teams/{}/roster'.format(team_id))
        resp.raise_for_status()
        roster = resp.json().get('roster')
        roster_dict = {}
        for player in roster:
            roster_dict[player.get('person').get('id')] = self.json_to_obj(player)
        return roster_dict

    def get_player_stats(self, player_id):
        resp = requests.get(self._API + 'people/{}/stats?stats=yearByYear'.format(player_id))
        resp.raise_for_status()
        stats = self.json_to_obj(resp.json().get('stats')[0])
        return stats

    @staticmethod
    def json_to_obj(data):
        temp = json.dumps(data)
        # https://stackoverflow.com/questions/6578986/how-to-convert-json-data-into-a-python-object
        return json.loads(temp, object_hook=lambda d: namedtuple('Object', d.keys())(*d.values()))

    def get_nhl_id(self):
        return self._NHL_ID

    def get_nhl_teams(self):
        return self._Teams

File: js/HockeyAPI.js
function HockeyAPI() {
    this._URL = 'https://statsapi.web.nhl.com/api/v1/';
    this._Request = new XMLHttpRequest();
    this._NHL = 133;
}

/** @property {object} teams*/
HockeyAPI.prototype.getTeams = function(set_default) {
        this._Request.onload = function () {
            let teams = JSON.parse(this.responseText).teams;
            let teamSelector = document.getElementById('teamSelector');
            teams.sort(HockeyAPI.compareNames);
            teams.forEach(function (team) {
                let option = document.createElement('option');
                option.value = team.id;
                option.textContent = team.name;
                // default team selected set to:
                // penguins === 5
                // capitals === 15
                if (set_default && (team.id === 15)) {
                    option.selected = true;
                }
                teamSelector.appendChild(option);
                // console.log(team.name);
            });
        };
        this._Request.open('get', this._URL+'/teams', false);
        this._Request.send();
    };

/** @property {object} roster
 *  @property {object} person
 *  @property {string} fullName
 * */
HockeyAPI.prototype.getRoster = function(teamId, position, set_default) {
        this._Request.onload = function () {
            let roster = JSON.parse(this.responseText).roster;
            let playerSelector = document.getElementById('playerSelector');
            let positionCode = (position.localeCompare("2")===0) ? "G" : "S";
            roster.sort(HockeyAPI.comparefullNames);
            roster.forEach(function (player) {
                if ((positionCode.localeCompare("G")===0) && (player.position.code.localeCompare("G")===0) ) {
                    HockeyAPI.createPlayerOption(player, set_default, playerSelector);
                }
                else if ((positionCode.localeCompare("S")===0) && (player.position.code.localeCompare("G")!==0) ) {
                    HockeyAPI.createPlayerOption(player, set_default, playerSelector);
                }

            });
        };
        this._Request.open('get', this._URL+'/teams/'+teamId+'/roster', false);
        this._Request.send();
    };



    // https://stackoverflow.com/questions/5362462/how-can-i-make-xhr-onreadystatechange-return-its-result
HockeyAPI.prototype.getPlayerStats = function(playerId, jsonResponse) {
        this._Request.onload = function () {
            jsonResponse.setResponse(JSON.parse(this.responseText).stats[0]);
        };
        this._Request.open('get', this._URL+'people/'+playerId+'/stats?stats=yearByYear', false);
        this._Request.send();
    };

/**
 * @property {object} people
 * */
HockeyAPI.prototype.getPlayerInformation = function(playerId, jsonResponse) {
    this._Request.onload = function () {
        jsonResponse.setResponse(JSON.parse(this.responseText).people[0]);
    };
    this._Request.open('get', this._URL+'people/'+playerId, false);
    this._Request.send();
};

HockeyAPI.createPlayerOption = function(player, set_default, playerSelector) {
    // console.log(player.fullName);
    let option = document.createElement('option');
    option.value = player.person.id;
    option.textContent = player.person.fullName;
    // default player selection set to:
    // crosby === 8471675
    // ovi === 8471214
    if (set_default && (player.person.id === 8471214)) {
        option.selected = true;
    }
    playerSelector.appendChild(option);
};

HockeyAPI.compareNames = function(a, b) {
        if (a.name < b.name)
            return -1;
        if (a.name > b.name)
            return 1;
        return 0;
    };

HockeyAPI.comparefullNames = function(a, b) {
        let aName = a.person.fullName.split(" ")[1];
        let bName = b.person.fullName.split(" ")[1];
        if (aName < bName)
            return -1;
        if (aName > bName)
            return 1;
        return 0;
    };

HockeyAPI.prototype.getNHLId = function() {
        return this._NHL;
    };
